README
                 Net::DBus
                 =========

Net::DBus provides a Perl XS API to the dbus inter-application
messaging system. The Perl API covers the core base level 
of the dbus APIs, not concerning itself yet with the GLib
or QT wrappers. For more information on dbus visit the 
project website at:

  http://www.freedesktop.org/software/dbus/

This version operates against DBus 1.00, and later

INSTALLATION
------------

To install this module type the following:

   perl Makefile.PL
   make
   make test
   sudo make install

The makefile calls the C<pkg-config> script to determine the
correct flags to pass to the compiler & linkers when building
the XS part of the module. Thus, ensure pkg-config is in
a directory listed by the $PATH environment. 

The pkg-config program will likely find the DBus install in /usr
provided by the base OS distribution, so if testing against an 
alternate  install of DBus, set the C$PKG_CONFIG_PATH env 
variable before generating the Makefile. For example, if your 
dbus installation is in C<$HOME/usr/dbus-cvs-tip> , then to build
and install against this version of DBus do

   export PKG_CONFIG_PATH=$HOME/usr/dbus-cvs-tip/lib/pkg-config
   perl Makefile.PL PREFIX=$HOME/usr/dbus-cvs-tip
   make
   make test
   make install


DEPENDENCIES
------------

In keeping with the C API, the Perl DBus implementation
has minimal external dependancies:

  Time::HiRes
  XML::Twig
  XML::Parser

And to run the test suite:

  Test::More
  Test::Pod (optional, but recommended)
  Test::Pod::Coverage (optional, but recommended)

Although they are not part of the base Perl distribution, most 
OS distributor will already provide addon packages containing
these modules. Failing this, they are all available from CPAN
http://search.cpan.org/

EXAMPLES
--------

There are a number of example programs in the examples/
subdirectory demonstrating use of the high level application
developer APIs.

  example-service.pl         - Providing a simple service
  example-client.pl          - Talking to a remote service
  example-signal-emitter.pl  - How to broadcast signals
  example-signal-receiver.pl - How to receive signals
  dump-object.pl             - Dump info about dbus objects
  lshal.pl                   - Dump names of all HAL devices

CONTRIBUTIONS
-------------

Contributions both simple bug fixes & new features are
always welcome. Please supply patches in context, or 
unified diff format. A simple method to generate such a 
patch is as follows:

  * Clean out generated files from your working
    directory:

        make distclean

  * Rename your working directory to have '-new' 
    extension: 

        mv DBus-1.0.0 DBus-1.0.0-new

  * Extract a pristine copy of the source:

        gunzip -c DBus-1.0.0.tar.gz | tar xf -
        mv DBus-1.0.0 DBus-1.0.0-orig

  * Generate the patch:

        diff -ruN DBus-1.0.0-orig DBus-1.0.0-new \
          > DBus-1.0.0-[something].patch
        gzip DBus-1.0.0-[something].patch


Send the resulting to .patch.gz file directly to 
Daniel Berrange <dan at berrange dot com>

COPYRIGHT AND LICENCE
---------------------

Copyright (C) 2004-2011 Daniel Berrange

Net-DBus may be redistributed and/or modified under the terms of Perl itself.
Either:

a) the GNU General Public License as published by the Free
   Software Foundation; either version 2, or (at your option) any
   later version

or

b) the "Artistic License"


See the file "COPYING" for the full text of each license

-- End



CHANGES

New in 1.0.0:

 - Updated to require minimum dbus >= 1.0.0
 - Automatically track change in ownership of bus names
   for signal handlers
 - Strict validation of method invocation against introspection
   data on exported objects
 - Improved error messages for invalid interfaces
 - Add API for disconnecting an object from a signal
 - Implement GetAll methods on properties interface
 - Allow leading _ in interface names
 - Other minor fixes

Changes since 0.33.5

 - Fix introspection XML handling when exporting objects with child
   objects

 - Improve output of Net::DBus::Dumper

 - Add support for providing parameter & return value names in
   introspection XML

 - Fixes to marshalling of variants

 - Fix handling of compound data types within object properties

 - Remove non-portable makefile rules

 - Fix ref counting bugs in error path.

Changes since 0.33.4

 - Added support for getting private bus connections for apps which
   don't want to deal with a shared bus

 - Fix test case to use a private connection

 - On Perl builds where integers are 32-bits, the DBus 64 bit integer
   types will be serialized to/from the Perl String type instead of
   calling 'die'.

 - Fix signature when marshalling dicts on newer DBus builds

 - Fix calling of disconnect wrt to newer DBus semantics

 - Make introspection much more tolerant of missing information
   about methods/properties/signals.

 - Fix use of magic values & added tests

 - Export the Net::DBus::Dumper methods correctly.

Changes since 0.33.3

 - Fixed service owner used for org.freedesktop.DBus object
   to make signal handling on the bus work again

 - Pass return value for signal handling callbacks all the
   way back to DBus

 - Fix multiple problems with marshalling of variant data
   types

 - Replace use of dbus_connection_disconnect with _close
   when compiling against dbus >= 0.90

 - Call dbus_connection_unref in the DESTROY method of
   connection object

 - Fix reference counting in connection & pending call
   objects

 - Added example of galago desktop notifications

 - Fix test suite errors

 - Added missing import statement

 - Throw Net::DBus::Error if an async call fails

Changes since 0.33.2

 - Fixed parsing of introspection data if there are processing
   instructions, or other non-data nodes before the root element.

 - Replace use of XML::Grove with XML::Twig when parsing the 
   introspection XML documents, since the former has not had any
   updates / bug fixes since 1999(!), and several people have 
   reported problems using it on Perl 5.8.x 

 - Made all Perl scripts / modules / tests use 'strict' and
   'warnings' pragmas

 - Turn Net::DBus::Error into fully fledged object which services
   can sub-class to allow explicit error handling by clients.

 - In _dispatch method of Net::DBus::Object ensure that any 
   instances of Net::DBus::Error thrown by the method call
   are explicitly serialized into DBus errors, rather than 
   a generic 'org.freedesktop.DBus.Failed'.

 - Change re-distribution license from GPL, to GPL / Perl Artistic,
   matching the terms of Perl itself.

 - Add support for registering a callback on Net::DBus::ASyncReply
   objects to allow notification of completion for asynchronous
   method calls

Changes since 0.33.1

 - Fixed handling of variants in introspection data

 - Added binding for the DBusPendingCall C object

 - Added some missing RPM dependancies on XML libs, and on minimum
   required dbus version

 - Added support for doing asynchronous method calls, and fire-and-
   forgot calls for methods whose return status is not desired. Use
   the constants in Net::DBus::Annotation module to indicate desired
   call mode. Default is to do synchronous blocking calls.

 - Added support for the 16-bit integer, signature and object path
   data types

 - Made introspection of root objects compliant with upsteam spec,
   by calling introspect on the root object, "/", rather than a 
   Perl specific magic object path.

Changes since 0.32.3

 - Constructor for Net::DBus::Object allows another Net::DBus::Object
   to be passed instead of the Net::DBus::Service, to create child
   objects specifying only a relative path.

 - Updated minimum required DBus version to be 0.33 to gain access
   to the unregister_object_path API

 - Add a disconnect() method to Net::DBus::Object to make it possible 
   to unregister object from the bus & thus make it possible to destroy 
   objects which are no longer required / relevant.

 - Unregister all child objects if we are unregistered ourselves

 - Fix numerous POD errors identified by Test::Pod and podchecker

 - Increase POD documentation to get 100% coverage of all APIs,
   verified by Test::Pod::Coverage

Changes since 0.32.2

 - Introspection data is used only as hint, so if an object
   exports many methods, but only provides partial introspection
   data, remote calls fallback to regular typing rules

 - Re-add dbus_XXX convenience methods to Net::DBus to allow
   clients to do explicit type casting. Must be requested at
   export time, using 'Net::DBus qw(:typing)'.

 - Update all example programs to run against session bus,
   since there are no security rules to enable them to work
   on system bus.

 - Print out warning upon use, if a method, signal, or property
   is annotated with the 'org.freedesktop.DBus.Deprecated' flag.

 - Do not wait for a method reply if the method is annotated
   with the 'org.freedesktop.DBus.Method.NoReply' flag.

 - Extend Net::DBus::Exporter to enable methods, signals, and
   properties to be annotated.

 - Add support for 'org.freedesktop.DBus.Method.NoReply' and
   'org.freedesktop.DBus.Deprecated' annotations when exporting
   objects

 - Add a pure in-memory bus implementation for facilitating
   creation of unit tests which would otherwise require making
   a connection to a 'live' message bus. Can be accessed via:
 
      Net::DBus->test


lib/Net/DBus.pm
# -*- perl -*-
#
# Copyright (C) 2004-2006 Daniel P. Berrange
#
# This program is free software; You can redistribute it and/or modify
# it under the same terms as Perl itself. Either:
#
# a) the GNU General Public License as published by the Free
#   Software Foundation; either version 2, or (at your option) any
#   later version,
#
# or
#
# b) the "Artistic License"
#
# The file "COPYING" distributed along with this file provides full
# details of the terms and conditions of the two licenses.

=pod

=head1 NAME

Net::DBus - Perl extension for the DBus message system

=head1 SYNOPSIS


  ####### Attaching to the bus ###########

  use Net::DBus;

  # Find the most appropriate bus
  my $bus = Net::DBus->find;

  # ... or explicitly go for the session bus
  my $bus = Net::DBus->session;

  # .... or explicitly go for the system bus
  my $bus = Net::DBus->system


  ######## Accessing remote services #########

  # Get a handle to the HAL service
  my $hal = $bus->get_service("org.freedesktop.Hal");

  # Get the device manager
  my $manager = $hal->get_object("/org/freedesktop/Hal/Manager",
				 "org.freedesktop.Hal.Manager");

  # List devices
  foreach my $dev (@{$manager->GetAllDevices}) {
      print $dev, "\n";
  }


  ######### Providing services ##############

  # Register a service known as 'org.example.Jukebox'
  my $service = $bus->export_service("org.example.Jukebox");


=head1 DESCRIPTION

Net::DBus provides a Perl API for the DBus message system.
The DBus Perl interface is currently operating against
the 0.32 development version of DBus, but should work with
later versions too, providing the API changes have not been
too drastic.

Users of this package are either typically, service providers
in which case the L<Net::DBus::Service> and L<Net::DBus::Object>
modules are of most relevance, or are client consumers, in which
case L<Net::DBus::RemoteService> and L<Net::DBus::RemoteObject>
are of most relevance.

=head1 METHODS

=over 4

=cut

package Net::DBus;

use 5.006;
use strict;
use warnings;

BEGIN {
    our $VERSION = '0.33.5';
    require XSLoader;
    XSLoader::load('Net::DBus', $VERSION);
}

use Net::DBus::Binding::Bus;
use Net::DBus::Service;
use Net::DBus::RemoteService;
use Net::DBus::Test::MockConnection;
use Net::DBus::Binding::Value;

use vars qw($bus_system $bus_session);

use Exporter qw(import);

use vars qw(@EXPORT_OK %EXPORT_TAGS);

@EXPORT_OK = qw(dbus_int16 dbus_uint16 dbus_int32 dbus_uint32 dbus_int64 dbus_uint64
		dbus_byte dbus_boolean dbus_string dbus_double
		dbus_object_path dbus_signature
		dbus_struct dbus_array dbus_dict dbus_variant);

%EXPORT_TAGS = (typing => [qw(dbus_int16 dbus_uint16 dbus_int32 dbus_uint32 dbus_int64 dbus_uint64
			      dbus_byte dbus_boolean dbus_string dbus_double
			      dbus_object_path dbus_signature
			      dbus_struct dbus_array dbus_dict dbus_variant)]);

=item my $bus = Net::DBus->find(%params);

Search for the most appropriate bus to connect to and
return a connection to it. The heuristic used for the
search is

  - If DBUS_STARTER_BUS_TYPE is set to 'session' attach
    to the session bus

  - Else If DBUS_STARTER_BUS_TYPE is set to 'system' attach
    to the system bus

  - Else If DBUS_SESSION_BUS_ADDRESS is set attach to the
    session bus

  - Else attach to the system bus

The optional C<params> hash can contain be used to specify
connection options. The only support option at this time
is C<nomainloop> which prevents the bus from being automatically
attached to the main L<Net::DBus::Reactor> event loop.

=cut

sub find {
    my $class = shift;

    if ($ENV{DBUS_STARTER_BUS_TYPE} &&
	$ENV{DBUS_STARTER_BUS_TYPE} eq "session") {
	return $class->session(@_);
    } elsif ($ENV{DBUS_STARTER_BUS_TYPE} &&
	     $ENV{DBUS_STARTER_BUS_TYPE} eq "system") {
	return $class->system(@_);
    } elsif (exists $ENV{DBUS_SESSION_BUS_ADDRESS}) {
	return $class->session(@_);
    } else {
	return $class->system;
    }
}

=item my $bus = Net::DBus->system(%params);

Return a handle for the system message bus. Note that the
system message bus is locked down by default, so unless appropriate
access control rules are added in /etc/dbus/system.d/, an application
may access services, but won't be able to export services.
The optional C<params> hash can contain be used to specify
connection options. The only support option at this time
is C<nomainloop> which prevents the bus from being automatically
attached to the main L<Net::DBus::Reactor> event loop.

=cut

sub system {
    my $class = shift;
    my %params = @_;
    if ($params{private}) {
	return $class->_new(Net::DBus::Binding::Bus->new(type => &Net::DBus::Binding::Bus::SYSTEM, private => 1), @_);
    }

    unless ($bus_system) {
	$bus_system = $class->_new(Net::DBus::Binding::Bus->new(type => &Net::DBus::Binding::Bus::SYSTEM), @_);
    }
    return $bus_system
}

=item my $bus = Net::DBus->session(%params);

Return a handle for the session message bus.
The optional C<params> hash can contain be used to specify
connection options. The only support option at this time
is C<nomainloop> which prevents the bus from being automatically
attached to the main L<Net::DBus::Reactor> event loop.

=cut

sub session {
    my $class = shift;
    my %params = @_;
    if ($params{private}) {
	return $class->_new(Net::DBus::Binding::Bus->new(type => &Net::DBus::Binding::Bus::SESSION, private => 1), @_);
    }

    unless ($bus_session) {
	$bus_session = $class->_new(Net::DBus::Binding::Bus->new(type => &Net::DBus::Binding::Bus::SESSION), @_);
    }
    return $bus_session;
}


=item my $bus = Net::DBus->test(%params);

Returns a handle for a virtual bus for use in unit tests. This bus does
not make any network connections, but rather has an in-memory message
pipeline. Consult L<Net::DBus::Test::MockConnection> for further details
of how to use this special bus.

=cut

# NB. explicitly do *NOT* cache, since unit tests
# should always have pristine state
sub test {
    my $class = shift;
    return $class->_new(Net::DBus::Test::MockConnection->new());
}

=item my $bus = Net::DBus->new($address, %params);

Return a connection to a specific message bus.  The C<$address>
parameter must contain the address of the message bus to connect
to. An example address for a session bus might look like
C<unix:abstract=/tmp/dbus-PBFyyuUiVb,guid=191e0a43c3efc222e0818be556d67500>,
while one for a system bus would look like C<unix:/var/run/dbus/system_bus_socket>.
The optional C<params> hash can contain be used to specify
connection options. The only support option at this time
is C<nomainloop> which prevents the bus from being automatically
attached to the main L<Net::DBus::Reactor> event loop.

=cut

sub new {
    my $class = shift;
    my $nomainloop = shift;
    return $class->_new(Net::DBus::Binding::Bus->new(address => shift), @_);
}

sub _new {
    my $class = shift;
    my $self = {};

    $self->{connection} = shift;
    $self->{signals} = [];
    $self->{services} = {};

    my %params = @_;

    bless $self, $class;

    unless ($params{nomainloop}) {
	if (exists $INC{'Net/DBus/Reactor.pm'}) {
	    my $reactor = $params{reactor} ? $params{reactor} : Net::DBus::Reactor->main;
	    $reactor->manage($self->get_connection);
	}
	# ... Add support for GLib and POE
    }

    $self->get_connection->add_filter(sub { return $self->_signal_func(@_); });

    $self->{bus} = Net::DBus::RemoteService->new($self, "org.freedesktop.DBus", "org.freedesktop.DBus");

    return $self;
}

=item my $connection = $bus->get_connection;

Return a handle to the underlying, low level connection object
associated with this bus. The returned object will be an instance
of the L<Net::DBus::Binding::Bus> class. This method is not intended
for use by (most!) application developers, so if you don't understand
what this is for, then you don't need to be calling it!

=cut

sub get_connection {
    my $self = shift;
    return $self->{connection};
}

=item my $service = $bus->get_service($name);

Retrieves a handle for the remote service identified by the
service name C<$name>. The returned object will be an instance
of the L<Net::DBus::RemoteService> class.

=cut

sub get_service {
    my $self = shift;
    my $name = shift;

    if ($name eq "org.freedesktop.DBus") {
	return $self->{bus};
    }

    my $owner = $name;
    if ($owner !~ /^:/) {
	$owner = $self->get_service_owner($name);
	if (!$owner) {
	    $self->get_bus_object->StartServiceByName($name, 0);
	    $owner = $self->get_service_owner($name);
	}
    }

    unless (exists $self->{services}->{$owner}) {
	$self->{services}->{$owner} = Net::DBus::RemoteService->new($self, $owner, $name);
    }
    return $self->{services}->{$owner};
}

=item my $service = $bus->export_service($name);

Registers a service with the bus, returning a handle to
the service. The returned object is an instance of the
L<Net::DBus::Service> class.

=cut

sub export_service {
    my $self = shift;
    my $name = shift;
    return Net::DBus::Service->new($self, $name);
}

=item my $object = $bus->get_bus_object;

Retrieves a handle to the bus object, C</org/freedesktop/DBus>,
provided by the service C<org.freedesktop.DBus>. The returned
object is an instance of L<Net::DBus::RemoteObject>

=cut

sub get_bus_object {
    my $self = shift;

    my $service = $self->get_service("org.freedesktop.DBus");
    return $service->get_object('/org/freedesktop/DBus',
				'org.freedesktop.DBus');
}


=item my $name = $bus->get_unique_name;

Retrieves the unique name of this client's connection to
the bus.

=cut

sub get_unique_name {
    my $self = shift;

    return $self->get_connection->get_unique_name
}

=item my $name = $bus->get_service_owner($service);

Retrieves the unique name of the client on the bus owning
the service named by the C<$service> parameter.

=cut

sub get_service_owner {
    my $self = shift;
    my $service = shift;

    my $bus = $self->get_bus_object;
    my $owner = eval {
	$bus->GetNameOwner($service);
    };
    if ($@) {
	if (UNIVERSAL::isa($@, "Net::DBus::Error") &&
	    $@->{name} eq "org.freedesktop.DBus.Error.NameHasNoOwner") {
	    $owner = undef;
	} else {
	    die $@;
	}
    }
    return $owner;
}


sub _add_signal_receiver {
    my $self = shift;
    my $receiver = shift;
    my $signal_name = shift;
    my $interface = shift;
    my $service = shift;
    my $path = shift;

    my $rule = $self->_match_rule($signal_name, $interface, $service, $path);

    push @{$self->{signals}}, [$receiver, $rule, $signal_name, $interface, $service, $path];
    $self->{connection}->add_match($rule);
}

sub _remove_signal_receiver {
    my $self = shift;
    my $receiver = shift;
    my $signal_name = shift;
    my $interface = shift;
    my $service = shift;
    my $path = shift;

    my $rule = $self->_match_rule($signal_name, $interface, $service, $path);

    my @signals;
    foreach (@{$self->{signals}}) {
	if ($_->[0] eq $receiver &&
	    defined $_->[1] &&
	    $_->[1] eq $rule) {
	    $self->{connection}->remove_match($rule);
	} else {
	    push @signals, $_;
	}
    }
    $self->{signals} = \@signals;
}


sub _match_rule {
    my $self = shift;
    my $signal_name = shift;
    my $interface = shift;
    my $service = shift;
    my $path = shift;

    my $rule = "type='signal'";
    if ($interface) {
	$rule .= ",interface='$interface'";
    }
    if ($service) {
	if ($service !~ /^:/) {
	    # Resolve service name to a client id
	    $service = $self->get_service_owner($service);
	}
	if ($service) {
	    $rule .= ",sender='$service'";
	}
    }
    if ($path) {
	$rule .= ",path='$path'";
    }
    if ($signal_name) {
	$rule .= ",member='$signal_name'";
    }
    return $rule;
}


sub _rule_matches {
    my $self = shift;
    my $rule = shift;
    my $member = shift;
    my $interface = shift;
    my $sender = shift;
    my $path = shift;

    my %bits;
    map {
	if (/^(\w+)='(.*)'$/) {
	    $bits{$1} = $2;
	}
    } split /,/, $rule;


    if (exists $bits{member} &&
	$bits{member} ne $member) {
	return 0;
    }
    if (exists $bits{interface} &&
	$bits{interface} ne $interface) {
	return 0;
    }
    if (exists $bits{sender} &&
	$bits{sender} ne $sender) {
	return 0;
    }
    if (exists $bits{path} &&
	$bits{path} ne $path) {
	return 0;
    }
    return 1;
}

sub _signal_func {
    my $self = shift;
    my $connection = shift;
    my $message = shift;

    return 0 unless $message->get_type() == &Net::DBus::Binding::Message::MESSAGE_TYPE_SIGNAL;

    my $interface = $message->get_interface;
    my $sender = $message->get_sender;
    my $path = $message->get_path;
    my $member = $message->get_member;

    my $handled = 0;
    foreach my $handler (grep { defined $_->[1] &&
				$self->_rule_matches($_->[1], $member, $interface, $sender, $path) }
			 @{$self->{signals}}) {
	my $callback = $handler->[0];
	&$callback($message);
	$handled = 1;
    }

    return $handled;
}

=back

=head1 DATA TYPING METHODS

These methods are not usually used, since most services provide introspection
data to inform clients of their data typing requirements. If introspection data
is incomplete, however, it may be neccessary for a client to mark values with
specific data types. In such a case, the following methods can be used. They
are not, however, exported by default so must be requested at import time by
specifying 'use Net::DBus qw(:typing)'

=over 4

=item $typed_value = dbus_int16($value);

Mark a value as being a signed, 16-bit integer.

=cut

sub dbus_int16 {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_INT16,
					  $_[0]);

}

=item $typed_value = dbus_uint16($value);

Mark a value as being an unsigned, 16-bit integer.

=cut


sub dbus_uint16 {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_UINT16,
					  $_[0]);
}

=item $typed_value = dbus_int32($value);

Mark a value as being a signed, 32-bit integer.

=cut

sub dbus_int32 {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_INT32,
					  $_[0]);

}

=item $typed_value = dbus_uint32($value);

Mark a value as being an unsigned, 32-bit integer.

=cut


sub dbus_uint32 {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_UINT32,
					  $_[0]);
}

=item $typed_value = dbus_int64($value);

Mark a value as being an unsigned, 64-bit integer.

=cut



sub dbus_int64 {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_INT64,
					  $_[0]);

}

=item $typed_value = dbus_uint64($value);

Mark a value as being an unsigned, 64-bit integer.

=cut



sub dbus_uint64 {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_UINT64,
					  $_[0]);
}

=item $typed_value = dbus_double($value);

Mark a value as being a double precision IEEE floating point.

=cut



sub dbus_double {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_DOUBLE,
					  $_[0]);
}

=item $typed_value = dbus_byte($value);

Mark a value as being an unsigned, byte.

=cut



sub dbus_byte {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_BYTE,
					  $_[0]);
}

=item $typed_value = dbus_string($value);

Mark a value as being a UTF-8 string. This is not usually required
since 'string' is the default data type for any Perl scalar value.

=cut



sub dbus_string {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_STRING,
					  $_[0]);
}

=item $typed_value = dbus_signature($value);

Mark a value as being a UTF-8 string, whose contents is a valid
type signature

=cut



sub dbus_signature {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_SIGNATURE,
					  $_[0]);
}

=item $typed_value = dbus_object_path($value);

Mark a value as being a UTF-8 string, whose contents is a valid
object path.

=cut

sub dbus_object_path {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_OBJECT_PATH,
					  $_[0]);
}

=item $typed_value = dbus_boolean($value);

Mark a value as being an boolean

=cut



sub dbus_boolean {
    return Net::DBus::Binding::Value->new(&Net::DBus::Binding::Message::TYPE_BOOLEAN,
					  $_[0]);
}

=item $typed_value = dbus_array($value);

Mark a value as being an array

=cut


sub dbus_array {
    return Net::DBus::Binding::Value->new([&Net::DBus::Binding::Message::TYPE_ARRAY],
					  $_[0]);
}

=item $typed_value = dbus_struct($value);

Mark a value as being a structure

=cut


sub dbus_struct {
    return Net::DBus::Binding::Value->new([&Net::DBus::Binding::Message::TYPE_STRUCT],
					  $_[0]);
}

=item $typed_value = dbus_dict($value);

Mark a value as being a dictionary

=cut

sub dbus_dict{
    return Net::DBus::Binding::Value->new([&Net::DBus::Binding::Message::TYPE_DICT_ENTRY],
					  $_[0]);
}

=item $typed_value = dbus_variant($value);

Mark a value as being a variant

=cut

sub dbus_variant{
    return Net::DBus::Binding::Value->new([&Net::DBus::Binding::Message::TYPE_VARIANT],
					  $_[0]);
}

=pod

=back

=head1 SEE ALSO

L<Net::DBus>, L<Net::DBus::RemoteService>, L<Net::DBus::Service>,
L<Net::DBus::RemoteObject>, L<Net::DBus::Object>,
L<Net::DBus::Exporter>, L<Net::DBus::Dumper>, L<Net::DBus::Reactor>,
C<dbus-monitor(1)>, C<dbus-daemon-1(1)>, C<dbus-send(1)>, L<http://dbus.freedesktop.org>,

=head1 AUTHOR

Daniel Berrange <dan@berrange.com>

=head1 COPYRIGHT

Copyright 2004-2005 by Daniel Berrange

=cut

1;
